File: backend/requests-service/db/requests.js
/* connect db */
import db from './connection.js';

/* create table */
export const createTable = () => {
    const sql = `
        CREATE TABLE IF NOT EXISTS requests (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            name TEXT NOT NULL CHECK (length(trim(name))>0),
            email TEXT NOT NULL CHECK (length(trim(email))>0 AND email like '%_@_%._%'),
            comment TEXT NOT NULL CHECK (length(trim(comment))>0),
            created_at DATETIME DEFAULT CURRENT_TIMESTAMP
        );
    `;
    db.prepare(sql).run();
};

/* add request */
export const insertRequests = ({name, email, comment}) => {
    const sql = `
        INSERT INTO requests (name, email, comment)
        VALUES (?, ?, ?)
    `;
    db.prepare(sql).run(name, email, comment);
};

/* get all requests */
export const getRequests = () => {
    const sql = `
        SELECT * FROM requests
    `;
    const rows = db.prepare(sql).all();
    return rows;
};

/* get requets by id */
export const getRequestById = (id) => {
    const sql = `
        SELECT * from requests
        WHERE id = ?
    `;
    const row = db.prepare(sql).get(id);
    return row;
};

File: backend/models/requests.go
package models

import (
	"database/sql"
)

type Request struct {
	ID        int
	Name      string
	Email     string
	Comment   string
	CreatedAt string
}

func InsertRequest(database *sql.DB, name string, email string, comment string) error {
	_, err := database.Exec(
		`INSERT INTO requests (name, email, comment)
		VALUES (?, ?, ?);`, name, email, comment,
	)

	return err
}

func GetRequests(database *sql.DB) ([]Request, error) {
	rows, err := database.Query(`SELECT * FROM requests;`)

	if err != nil {
		return nil, err
	}

	defer rows.Close()

	var requests []Request

	for rows.Next() {
		var req Request

		err := rows.Scan(
			&req.ID,
			&req.Name,
			&req.Email,
			&req.Comment,
			&req.CreatedAt,
		)

		if err != nil {
			return nil, err
		}

		requests = append(requests, req)
	}

	if err = rows.Err(); err != nil {
		return nil, err
	}

	return requests, nil
}

func GetRequestById(database *sql.DB, id int) (Request, error) {
	var req Request

	err := database.QueryRow(
		`SELECT * FROM requests
		WHERE ID = ?;`, id,
	).Scan(
		&req.ID,
		&req.Name,
		&req.Email,
		&req.Comment,
		&req.CreatedAt,
	)

	if err != nil {
		return req, err
	}

	return req, nil

}
